**Statistics**
-
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

In [285]:
from scipy.stats import ttest_1samp

One Sample
-

In [286]:
sample_scores = [4.2, 3.8, 3.5, 4.0, 4.5, 3.3, 4.2, 4.1, 3.9, 4.3, 
                 3.7, 4.0, 4.2, 3.8, 3.6, 4.1, 3.9, 4.4, 4.0, 4.2, 
                 3.7, 3.9, 4.0, 4.1, 4.3, 3.6, 3.9, 4.0, 4.2, 4.2, 
                 3.8, 3.5, 4.0, 4.5, 3.3, 4.2, 4.1, 3.9, 4.3, 3.7, 
                 4.0, 4.2, 3.8, 3.6, 4.1, 3.9, 4.4, 4.0, 4.2, 3.7, 
                 3.9, 4.0, 4.1, 4.3, 3.6, 3.9, 4.0, 4.2]

In [287]:
mean=3

ttest_1samp - Function
-

Returns t -statistic and p-value for one-sample t-test.



In [288]:
t, p_value = ttest_1samp(sample_scores,mean)
print("t : ",t)
print("p_value : ",p_value)

t :  27.0438538051636
p_value :  3.3377047399340263e-34


Two sample Test
-

In [289]:
a=[85,88,92,78,90]
b=[91,89,82,87,88]

In [290]:
from scipy.stats import ttest_ind

In [291]:
t, P_value = ttest_ind(a,b)
print("t : ",t)
print("P_value : ",P_value)

t :  -0.2790320425660663
P_value :  0.7872977318566334


Z - Test 
-

In [292]:
import random

# Generate sample weight loss data (in pounds)

random.seed(42)  # For reproducibility

weight_loss_data = [random.uniform(0.5, 10) for _ in range(100)]

In [293]:
weight_loss_data

[6.574554585349896,
 0.7376021746153358,
 3.112778524506633,
 2.6205020124138163,
 7.496476534558118,
 6.9286451305176575,
 8.975705893196032,
 1.3259189099794535,
 4.5082572870100694,
 0.7830735846616683,
 2.577060760634232,
 5.300875236981943,
 0.7520917119967044,
 2.3889576815231606,
 6.67390215890547,
 5.6769440657305585,
 2.594185909386619,
 6.098023996821133,
 8.189589338439353,
 0.5617382169415797,
 8.155282892411675,
 7.132324252388155,
 3.7323799069209227,
 1.9770552482119248,
 9.593524185964421,
 3.6976481785699544,
 1.3810855121114052,
 1.4188055799179082,
 8.551196480300868,
 6.2353972979854655,
 8.167718596106612,
 7.43245197359127,
 5.594166868819657,
 9.744599757804021,
 4.096076583479358,
 5.744385997095656,
 8.37934431040345,
 6.375937647460337,
 8.686215552952383,
 5.984845379939239,
 7.193432444041773,
 0.935331644728791,
 2.6650336186896952,
 3.249185654220018,
 1.2580237807744612,
 2.7115134204297866,
 1.4595135793924265,
 3.140749229545875,
 6.539002220511802,
 3.

2. Calculate the Test Statistic (Z-Score)

Sample Mean (X̄): Calculate the sample mean of weight loss data.
Population Standard Deviation (σ): Known as 2 pounds.
Sample Size (n): 100 participants.

In [294]:
import numpy as np

# Calculate sample mean and sample standard deviation

sample_mean = np.mean(weight_loss_data)
sample_std = np.std(weight_loss_data, ddof=1)  # Use Bessel's correction for sample std

# Calculate Z-score

population_mean = 0  # Hypothesized population mean
z_score = (sample_mean - population_mean) / (sample_std / np.sqrt(len(weight_loss_data)))

sample_mean, sample_std, z_score

(5.05590780970096, 2.8025373784562273, 18.04046521758078)

3. Determine the Critical Value

Significance Level (α): Typically, 0.05 (corresponding to 95% confidence level).
Degrees of Freedom (df): n – 1 (99 degrees of freedom for this example).

In [295]:
from scipy.stats import norm

alpha = 0.05

df = len(weight_loss_data) - 1

# Calculate the critical Z-value

critical_value = norm.ppf(1 - alpha)

critical_value

1.6448536269514722

In [296]:
critical_value , z_score

(1.6448536269514722, 18.04046521758078)

 Make a Decision

Compare the calculated Z-score with the critical value.

If Z-Score > Critical Value, reject the null hypothesis (Ha).
If Z-Score ≤ Critical Value, fail to reject the null hypothesis (H0).

In [297]:
if(z_score > critical_value):
    print("Reject the null hypothesis")
else:
    print("Fail to reject the null hypothesis")

Reject the null hypothesis


CHI - SQUARE TEST
-

Chi-Square Test for Independence:

Chi-Square Test: The chi-square test is a statistical test used to assess the independence (or association) between two categorical variables. It helps determine whether there is a significant relationship between the variables or if they are independent.

Independence: In the context of the chi-square test, independence means that the two categorical variables are not related, and changes in one variable do not affect the distribution of the other variable.

Using Python (Mathematically)


In [298]:
import pandas as pd
import random
from scipy.stats import chi2_contingency

# Generate random data for 200 voters
random.seed(42)  # For reproducibility

# Generate random gender data
gender = random.choices(['Male', 'Female'], k=200)

# Generate random voting preferences data
voting_preferences = random.choices(['Candidate A', 'Candidate B', 'Undecided'], k=200)

# Create a dataframe from the data
data = {
    'Gender': gender,
    'Voting_Preferences': voting_preferences
}
df = pd.DataFrame(data)

1. Hypotheses

You want to test whether there is a significant association between the gender of voters and their voting preferences.

Null Hypothesis (H0): Gender and voting preferences are independent.
Alternative Hypothesis (Ha): Gender and voting preferences are not independent; there is an association.

2. Create a Contingency Table

Create a contingency table that cross-tabulates the two categorical variables (gender and voting preferences):

In [299]:
contingency_table = pd.crosstab(df['Gender'], df['Voting_Preferences'])

contingency_table

Voting_Preferences  Candidate A  Candidate B  Undecided
Gender                                                 
Female                       31           26         40
Male                         40           29         34

3. Calculate the Chi-Square Statistic

Calculate the chi-square statistic, which measures the discrepancy between observed and expected frequencies.

In [300]:
from scipy.stats import chi2_contingency

# Calculate chi-square statistic, p-value, degrees of freedom, and expected frequencies
chi2, p, dof, expected = chi2_contingency(contingency_table)
chi2, p, dof, expected

(1.6124190977333461,
 0.4465474787793734,
 2,
 array([[34.435, 26.675, 35.89 ],
        [36.565, 28.325, 38.11 ]]))

Chi-Square Statistic (χ²): Approximately 0.1463

P-Value (p): Calculated based on data

Degrees of Freedom (df): 2

Expected Frequencies: Calculated values based on the data

4. Determine the Critical Value

Specify the desired level of significance (α) to determine the critical chi-square value. Let’s assume α = 0.05.

In [301]:
from scipy.stats import chi2

alpha = 0.05

critical_value = chi2.ppf(1 - alpha, dof)

critical_value

5.991464547107979

Critical Value (χ²_critical): Approximately 5.991 (for α = 0.05 and df = 2)

5. Make a Decision

Compare the calculated chi-square statistic with the critical value.

If χ² > χ²_critical, reject the null hypothesis.

If χ² ≤ χ²_critical, fail to reject the null hypothesis.

In [303]:
# Make a decision
chi2,critical_value


(<scipy.stats._continuous_distns.chi2_gen at 0x255efc98e90>, 5.991464547107979)

In [ ]:
# Make a decision

reject_null = chi2 > critical_value

reject_null

ANOVA
-

In [306]:
import numpy as np
from scipy.stats import f

# Given data
school_1 = np.array([85, 88, 90, 92, 86])
school_2 = np.array([78, 82, 80, 85, 88])
school_3 = np.array([92, 94, 89, 88, 90])

# Step 1: Calculate Grand Mean
all_scores = np.concatenate([school_1, school_2, school_3])
grand_mean = np.mean(all_scores)

# Step 2: Calculate SST
sst = np.sum((all_scores - grand_mean)**2)

# Step 3: Calculate SSB
ssb = np.sum([len(school) * (np.mean(school) - grand_mean)**2 for school in [school_1, school_2, school_3]])

# Step 4: Calculate SSW
ssw = np.sum([(score - np.mean(school))**2 for school, scores in zip([school_1, school_2, school_3], [school_1, school_2, school_3]) for score in scores])

# Step 5: Degrees of Freedom
df_between = 3 - 1
df_within = len(all_scores) - 3

# Step 6: Calculate Mean Squares
msb = ssb / df_between
msw = ssw / df_within

# Step 7: Calculate F-statistic
f_statistic = msb / msw

# Step 8: Compare with Critical Value or P-value
p_value = 1 - f.cdf(f_statistic, df_between, df_within)

# Print results
print(f"F-statistic: {f_statistic}")
print(f"P-value: {p_value}")

alpha = 0.05
if p_value < alpha:
    print("Reject the null hypothesis: \nThere are significant differences in test scores among the three schools.")
else:
    print("Fail to reject the null hypothesis: \nThere are no significant differences in test scores among the three schools.")

F-statistic: 8.483221476510069
P-value: 0.005054945419651569
Reject the null hypothesis: 
There are significant differences in test scores among the three schools.


Using Python Libraries

Perform the One-Way ANOVA
We’ll use the scipy.stats library to perform the one-way ANOVA.

In [307]:
from scipy.stats import f_oneway

# Group the data by school
groups = [df['Test_Score'][df['School'] == school] for school in df['School'].unique()]

# Perform one-way ANOVA
f_statistic, p_value = f_oneway(*groups)
f_statistic, p_value

KeyError: 'School'

In [308]:
alpha = 0.05  # Significance level

critical_value = 3.8853  # Critical F-value for α = 0.05 and df = (2, 12)

# Make a decision

reject_null = f_statistic > critical_value
reject_null

True

In [309]:
import scipy.stats as stats
import numpy as np

# Data: Test scores for three schools
school_1 = np.array([85, 88, 90, 92, 86])
school_2 = np.array([78, 82, 80, 85, 88])
school_3 = np.array([92, 94, 89, 88, 90])

# Perform one-way ANOVA
statistic, p_value = stats.f_oneway(school_1, school_2, school_3)

# Print results
print(f"One-Way ANOVA Statistic: {statistic}")
print(f"P-value: {p_value}")

# Check significance at a 0.05 significance level
alpha = 0.05
if p_value < alpha:
    print("\nReject the null hypothesis: There are significant differences in test scores among the three schools.")
else:
    print("\nFail to reject the null hypothesis: There are no significant differences in test scores among the three schools.")


One-Way ANOVA Statistic: 8.483221476510069
P-value: 0.00505494541965152

Reject the null hypothesis: There are significant differences in test scores among the three schools.


In [312]:
school_1 = [85, 88, 90, 92, 86]
school_2 = [78, 82, 80, 85, 88]
school_3 = [92, 94, 89, 88, 90]
scoresdata = {    'School': ['school_1']*5 + ['school_2']*5 + ['school_3']*5,    'Test_Score': school_1 + school_2 + school_3}

df = pd.DataFrame(data)
df

Gender Voting_Preferences
0    Female          Undecided
1      Male          Undecided
2      Male          Undecided
3      Male        Candidate A
4    Female        Candidate B
..      ...                ...
195  Female        Candidate A
196  Female        Candidate B
197    Male        Candidate B
198  Female        Candidate B
199  Female        Candidate B

[200 rows x 2 columns]